### PR TITLE
Refactoring containers and sagas

### PR DESCRIPTION
**Pull request by alejandropaciotti**
This PR include a refactor for the implementation of Containers and Screens.
Also include the separation of sagas, reducers, actions in each folder.

Related with [this](https://github.com/thecodingmachine/react-native-boilerplate/issues/6) issue

### CODE REVIEW

File: App/Containers/Root/index.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+          ref={(navigatorRef) => {
+            NavigationService.setTopLevelNavigator(navigatorRef)
+          }}
+        />
+      </RootScreen>
+    )
+  }
+}
+
+const mapStateToProps = (state) => ({})
```
Comment (jahglow) [inline App/Containers/Root/index.js]: why would you pass empty object as state? pass null instead

**Comment by isilher** [comment]
This looks great! In my opinion this was the only part really missing in the boilerplate.

If this gets accepted, the old example structure can be removed, right?

**Comment by alejandropaciotti** [comment]
> This looks great! In my opinion this was the only part really missing in the boilerplate.
> 
> If this gets accepted, the old example structure can be removed, right?

Yes !

**Comment by RobPando** [comment]
What do some of you think about renaming the `Containers` folder to `Features` folder along with this change? I personally do follow the containers/component conventions whole heartily. However, I feel like it does not make sense to name the folder `Containers` although the containers will live here. With this change (which I 100% agree and use in my daily work) these folders inside no longer are composed of only containers and they almost entirely resemble features in mobile development. These features as well, will almost certainly have components of their own which will be used only by that feature and creating a `Components` folder inside these features starts making sense, hence leaving the `App/Components` folder for common components that will be used by multiple features.

**Comment by alejandropaciotti** [comment]
> What do some of you think about renaming the `Containers` folder to `Features` folder along with this change? I personally do follow the containers/component conventions whole heartily. However, I feel like it does not make sense to name the folder `Containers` although the containers will live here. With this change (which I 100% agree and use in my daily work) these folders inside no longer are composed of only containers and they almost entirely resemble features in mobile development.
> 
> These features as well, will almost certainly have components of their own which will be used only by that feature and creating a `Components` folder inside these features starts making sense, hence leaving the `App/Components` folder for common components that will be used by multiple features.

I think the `containers` folder has to have only components connected to the state, that import actions and dispatch it.
I think that the aren't place for `components` inside the `containers` folder, because `components` are concerned in how the things look.

Just in case you don't have read, this are the post mencionated in the boilerplate:

https://medium.com/@dan_abramov/smart-and-dumb-components-7ca2f9a7c7d0

One paragraph related with this discuss says:

> I call them Container and Presentational components* but I also heard Fat and Skinny, Smart and Dumb, Stateful and Pure, Screens and Components, etc. These all are not exactly the same, but the core idea is similar.

**Comment by RobPando** [comment]
> I think the containers folder has to have only components connected to the state, that import actions and dispatch it.
I think that the aren't place for components inside the containers folder, because components are concerned in how the things look.

I think I suck at explaining, so I apologize for that, but like I said in my previous comment "I personally do follow the containers/component conventions whole heartily." I have read that article countless times and love it, so again, I follow that to the core. What I meant is that renaming the container's folder to features,  since you are now introducing actions and stores and reducers and sagas inside the containers folder they now start resembling features and the container component lives in the same folder now `App/Container/Example/index.js` I felt that the now `App/Containers` folder should be renamed to `App/Features` this way you could also have a `App/Features/Example/Components` folder where you would import into your container `App/Features/Example/index.js` because often time some containers have components that only **that** particular component will use. You can ignore my suggestion and comments if you understand what I mean and still disagree, just wanted to know what the thoughts was because if this gets merged then to me personally it no longer makes sense to me to have the `Containers` folder named that way since it is no longer composed of only container components. Additionally, if you disagree with the `Features` folder, where would you place components that are only used by a certain container? in a `Components/Example` folder? or just in the root `Components` folder? I feel like all this folder structure is a convention that the react-native community is missing.

**Comment by alejandropaciotti** [comment]
> > I think the containers folder has to have only components connected to the state, that import actions and dispatch it.
> > I think that the aren't place for components inside the containers folder, because components are concerned in how the things look.
> 
> I think I suck at explaining, so I apologize for that, but like I said in my previous comment "I personally do follow the containers/component conventions whole heartily." I have read that article countless times and love it, so again, I follow that to the core. What I meant is that renaming the container's folder to features, since you are now introducing actions and stores and reducers and sagas inside the containers folder they now start resembling features and the container component lives in the same folder now `App/Container/Example/index.js` I felt that the now `App/Containers` folder should be renamed to `App/Features` this way you could also have a `App/Features/Example/Components` folder where you would import into your container `App/Features/Example/index.js` because often time some containers have components that only **that** particular component will use. You can ignore my suggestion and comments if you understand what I mean and still disagree, just wanted to know what the thoughts was because if this gets merged then to me personally it no longer makes sense to me to have the `Containers` folder named that way since it is no longer composed of only container components. Additionally, if you disagree with the `Features` folder, where would you place components that are only used by a certain container? in a `Components/Example` folder? or just in the root `Components` folder? I feel like all this folder structure is a convention that the react-native community is missing.

I think that if you are making a component only for one container/feature you need to rethink that component. Because if you left in the container/feature/components you aren't not thinking in the reusability of that component.

I think you could have a container that had another containers, and that could replace the components inside containers/feature folder. I wouldn't put a component folder inside the containers folder. Because if thats components are connected with the state, then its should be a container. And if thats components aren't connected, then, there are only visual components. Then, just thinking in the reusability you need to put outside the containers folder.

Btw, all of that depends the requirements of the application, and this is only a boilerplate. You can use a name your folders as you want.

**Comment by RobPando** [comment]
> I wouldn't put a component folder inside the containers folder.

 I 100% agree. This is something that I would never do,  I did not suggest this in any of my comments and I believe we are on the same page on this.

> I think that if you are making a component only for one container/feature you need to rethink that component. Because if you left in the container/feature/components you aren't not thinking in the reusability of that component.

I disagree with this, I don't believe presentational components should be solely thought of as reusability, I think their purpose is not reusability alone, it will be of course for 90% of the time. But it can also be to slim down containers where it makes sense, you might have a unique custom footer or header for a particular component that fattens your container, then to me extracting it to its own personal component makes sense Or you might have a `FlatList` with a list of pressable links and to avoid arrow functions in props for `onPress` you would have a component that receives the link to better handle the press action such as

```javascript
// Features/Example/index.js or Features/Example/LinksList.js
...
renderItem({ item }) {
 return <LinksListItem onPress={this.openLink} link={item.link} />
}

render() {
<FlatList
  data={this.props.linkList}
  renderItem={this.renderItem}
 />
...

// Feature/Example/Components/LinksListItem.js
...
onPress() {
 const { onPress, link } = this.props
 onPress(link)
}

render() {
 // some styled component with touchable opacity and the onPress set to this.onPress
}

```


Aside from all that, you are right however, that this is only a boilerplate, and anyone can name it what they want, I just personally did not see this making sense if it is still being called `Containers`. In that case I see the current state making more sense where the store of example should be under its own store/redux folder as well as the sagas, the way it is now. Just my personal opinion, I do thank you for the nice discussion though 🙂👍

**Pull request closed without merging**